### PR TITLE
Clarify the behaviors of response addHeader and setHeader

### LINKED ISSUE #26483: Servlet 6.1: Clarify the behaviour of the various methods for adding headers when passed null or the empty string.
`https://github.com/jakartaee/servlet/issues/159`

### PR DESCRIPTION
**Pull request by pmd1nh**
fixes https://github.com/OpenLiberty/open-liberty/issues/26483

### CODE REVIEW

**Comment by pmd1nh** [comment]
#build        #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

File: dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/srt/SRTServletResponse61.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2023 IBM Corporation and others.
+ * Copyright (c) 2023-2024 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/srt/SRTServletResponse61.java]: Your copyright date ranges in this PR should be `2023, 2024` not `2023-2024`

**Comment by pmd1nh** [comment]
#build 
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat

**Comment by pmd1nh** [comment]
rebased; updated .classpath

#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat

**Review by pnicolucci — COMMENTED** [review]
Phu, looking at the changes for this issue in the Servlet community a couple of questions to ensure we're complete here:

setDateHeader
	1) If the header had already been set, the new value overwrites all previous values -> Is this how we already behave? Replacing all previous values and not just the first one found?

setHeader
	1) If the header had already been set, the new value overwrites all previous values. -> Same comment as setDateHeader
	2) This method has no effect if called after the response has been committed -> I think we're good here but it was added to the javadoc so I want to ensure that is the case!
	3) Passing {@code null} as the value removes all headers with the given name.  -> Is this our current beahvior?
	4) Note that the empty string is a valid header value. -> Is this our current behavior?


addHeader
	1) This method has no effect if called after the response has been committed. -> Is this our current behavior?
	2) Note that the empty string is a valid header value. -> Is this our current behavior?

setIntHeader
	1)  If the header had already been set, the new value overwrites all previous values. -> Is this our current behavior?

Can we add some additional tests to ensure we're covered here with the new JavaDoc?

**Comment by pmd1nh** [comment]
@pnicolucci I expanded the tests to cover the concerned areas.

setDateHeader:
1. Yes, new value will overwrite all previous values.  This has been our behavior.

setHeader:
1. Yes, overwriting all previous value for same header.
2. Yes, all of these headers are ignored if response has committed (no new test in this area).
3. Yes, remove the header if value is null
4. Yes

addHeader:
1. Yes,
2. Yes,

setIntHeader
1. Yes,

The only update from the current code is for addDateHeader:
1. can have multiple values for this add date header.  However, our server has not behaved like this.  I fixed it with this servlet 6.1 (See change in SRTServletResponse). Older servlet levels remain as-is (to avoid regression).
2. addHeader and addIntHeader behave correctly (i.e a header can have multiple values).  Tests are added to test these 3 API for multiple values.

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTServletResponse.java [hunk 2]
```diff
@@ -2252,7 +2252,12 @@ public void addDateHeader(String name, long date) {
                 throw new IllegalArgumentException("Cannot Set Content-Type to a Date value");
             }
             else {
-                setHeader(name, (createCompliantHttpDateString(date)));
+                if (com.ibm.ws.webcontainer.osgi.WebContainer.getServletContainerSpecLevel() < com.ibm.ws.webcontainer.osgi.WebContainer.SPEC_LEVEL_61) {
+                    setHeader(name, (createCompliantHttpDateString(date)));
+                }
+                else {
+                    addDateField(name, date);
+                }
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTServletResponse.java:2260]: Let's make sure to add this to the UFO as a Servlet 6.1 behavior difference for Open Liberty

File: dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/srt/SRTServletResponse61.java [hunk 3]
```diff
@@ -34,13 +34,97 @@ public void initForNextResponse(IResponse resp) {
...
+     * This method has no effect if {@code null} is passed for the {@code name} parameter.
+     * <p>
+     * Passing {@code null} as the value removes all headers with the given name.
      *
+     */
+    @Override
+    public void setHeader(String name, String value) {
+        if (name == null) {
+            if (TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINE)) {
+                logger.logp(Level.FINE, CLASS_NAME, "addIntHeader", "name is null; return. [" + this + "]");
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/srt/SRTServletResponse61.java]: trace should be "setHeader" not "addInitHeader" right?

File: dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/publish/servers/servlet61_AddAndSetHeaderTest/bootstrap.properties [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License 2.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-2.0/
+# 
+# SPDX-License-Identifier: EPL-2.0
+###############################################################################
+bootstrap.include=../testports.properties
+osgi.console=7777
+com.ibm.ws.logging.trace.specification=*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:GenericBNF=all:io.openliberty.webcontainer*=all
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/publish/servers/servlet61_AddAndSetHeaderTest/bootstrap.properties]: Do we need all of this tracing? Would you be comfortable cutting this down or removing it? We can always add it in if we have a build break and need further debug in a build.

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTServletResponse.java (hunk 2 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTServletResponse.java:2260]: Track via ToDo issue
 https://github.com/OpenLiberty/open-liberty/issues/27383

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Review by pnicolucci — APPROVED** [review]

**Comment by pmd1nh** [comment]
#libby

**Comment by pmd1nh** [comment]
1 unrelated failure

no component 90480 0 1
	com.ibm.ws.ui.tool.serverConfig_fat.gui FAT tests 5 0 1

**Pull request merged**